### ISSUE TITLE
Allow direct access to ICacheEntry and make it thread safe 

### ISSUE THREAD

**Issue by ohadschn** [aspnet/Caching#338]
Currently all methods that return a value from the cache (most importantly `TryGetValue`) directly return the underlying `object`. This is convenient but does not allow for a scenario of updating an entry's expiration after it has been added. Even if I retain a reference to the original entry returned from `CreateEntry` it won't be good enough as no memory barriers are raised when accessing its properties (so I might change the expiration time in my thread, but your `StartScanForExpiredItems` thread won't observe them).

**Comment by Tratcher** [comment]
Which expiration are you specifically trying to change and why? AbsoluteExpiration, AbsoluteExpirationRelativeToNow, SlidingExpiration, or ExpirationTokens?

**Comment by ohadschn** [comment]
I'm proposing to allow all types of expirations to be set after the value has been added to the cache. 
As for why, the entry's expiration date might be a function of the value. I want to be able to do this:
```cs
public Task<T> GetOrAddAsync<T>(
    string key, 
    Func<Task<T>> factory, 
    Action<T, ICacheEntry> expirationSetter, 
    LazyThreadSafetyMode mode)
{
    var lazy = (Lazy<Task<T>>) _cache.GetOrCreate(key, ce => new Lazy<Task<T>>(async () => 
        {
            var value = await factory();
            expirationSetter(value, ce);
            return value;
        }, mode));
    return lazy.Value;
}
```

For this to work well, I'd need https://github.com/aspnet/Caching/issues/240 too.

**Comment by ohadschn** [comment]
I think something like this could be broken too:

```cs
public Task<T> GetOrAddAsync<T>(
     string key, Func<Task<T>> factory, Action<ICacheEntry, T> expirationSetter)
{    
    return _cache.GetOrCreateAsync(key, async cacheEntry => 
    {
        var value = await factory().ConfigureAwait(false);
        expirationSetter(cacheEntry, value);
        return value;
    });
}
```
Because after the `await` you might be in a different thread whose changes are not observed by the `GetOrCreateAsync` thread that created the cache entry (and would later put it in the cache dictionary).

**Comment by Tratcher** [comment]
Have you tried using a custom IChangeToken that wrapped your value and managed its timeout internally?
https://github.com/aspnet/Common/blob/52c69ec4d9c0e3de40da092e3721ace0ce0ac489/src/Microsoft.Extensions.Primitives/IChangeToken.cs

**Comment by ohadschn** [comment]
Interesting, this could work:

```cs
public Task<T> GetOrAddAsync<T>(
    string key, Func<Task<T>> factory, Func<T, TimeSpan> expirationCalculator)
{    
    return _cache.GetOrCreateAsync(key, async cacheEntry => 
    {
        var cts = new CancellationTokenSource();
        cacheEntry.AddExpirationToken(new CancellationChangeToken(cts.Token));
        var value = await factory().ConfigureAwait(false);
        Task.Delay(expirationCalculator(value)).ContinueWith(t => cts.Cancel());
        return value;
    });
}
```
However this won't allow for sliding expirations. Also, https://github.com/aspnet/Caching/issues/240 would still be required to guarantee the factory is only called once.

**Comment by Tratcher** [comment]
Oh, I was thinking more along the lines of:
```
    return _cache.GetOrCreateAsync(key, async cacheEntry => 
    {
        var value = await factory().ConfigureAwait(false);
        var token = new CustomChangeToken(value);
        cacheEntry.AddExpirationToken(token);
        return value;
    });

public class CustomChangeToken : IChangeToken
{
  public CustomChangeToken(MyType value) { }
  public bool HasChanged { get => value.HasExpired }
  ...
}
```

Sliding expiration would only work if you kept pinging something on the value each time you used it. E.g. value.ExtendLifetime().

**Comment by ohadschn** [comment]
Yes, if you need custom expiration logic that might make more sense, but I reckon using the existing `CancellationChangeToken` rather than rolling your own would be simpler in most cases.

Regardless, I am a bit concerned about the thread safety of the code you posted. You operate on `cacheEntry` from a different thread than the one it was created in, so I don't believe changes made on it after the `await` are guaranteed to be observed by the creating thread without any memory barriers raised. Specifically you call `AddExpirationToken` which boils down to `List<t>.Add` which does not raise memory barriers as far as I can tell.

**Comment by Tratcher** [comment]
Fair point, I was assuming behavior of the existing synchronous extensions like GetOrAdd, not the async/lazy versions you've defined.
